File: plugins/modulation/tremolo/tremolo.js
// Tremolo — amplitude modulated by a sine LFO. Classic surf-rock /
// vintage amp "wobble". The LFO phase is held on the instance so the
// modulation continues smoothly across process() blocks within one
// render — without that the wobble would reset every 1024 samples and
// you'd hear a clicky chop instead of a glide.

function TremoloProcessor() {
  this.phase = 0;
}
TremoloProcessor.parameterDescriptors = [
  { name: 'rate',  label: 'Rate (Hz)', defaultValue: 5,    minValue: 0.1, maxValue: 20 },
  { name: 'depth', label: 'Depth',     defaultValue: 0.6,  minValue: 0,   maxValue: 1 }
];
TremoloProcessor.prototype.process = function(inputs, outputs, parameters) {
  var input = inputs[0][0];
  var output = outputs[0][0];
  if (!input || !output) return true;
  var rate = parameters.rate[0];
  var depth = parameters.depth[0];
  var phaseInc = (2 * Math.PI * rate) / sampleRate;
  for (var i = 0; i < input.length; i++) {
    // LFO range: 1 - depth .. 1 (never inverts polarity — preserves
    // formants and avoids the comb-filtery sound of through-zero mod).
    var lfo = 1 - depth * 0.5 * (1 - Math.cos(this.phase));
    output[i] = input[i] * lfo;
    this.phase += phaseInc;
    if (this.phase > 2 * Math.PI) this.phase -= 2 * Math.PI;
  }
  return true;
};

registerProcessor('tremolo', TremoloProcessor);

File: plugins/modulation/tremolo/Tremolo.java
package com.vocalmonitor.plugin.community;

import com.vocalmonitor.plugin.VocalMonitorNativePlugin;

// Tremolo (native port) — amplitude modulated by a sine LFO. The LFO
// range is 1-depth .. 1 so the wave never inverts polarity — classic
// surf-rock / vintage-amp wobble that preserves formants.
public final class Tremolo implements VocalMonitorNativePlugin {
    private float phase = 0f;
    private int sampleRate = 44100;
    private float rate = 5f, depth = 0.6f;

    @Override public void init(int sr) { this.sampleRate = sr; phase = 0f; }
    @Override public String[] parameterNames() { return new String[] { "rate", "depth" }; }
    @Override public float parameterMin(String n) { return "rate".equals(n) ? 0.1f : 0f; }
    @Override public float parameterMax(String n) { return "rate".equals(n) ? 20f : 1f; }
    @Override public float parameterDefault(String n) { return "rate".equals(n) ? 5f : 0.6f; }
    @Override public String parameterLabel(String n) { return "rate".equals(n) ? "Rate (Hz)" : "Depth"; }
    @Override public void setParameter(String n, float v) {
        if ("rate".equals(n)) rate = v;
        else if ("depth".equals(n)) depth = v;
    }

    @Override
    public void process(float[] input, float[] output) {
        final float phaseInc = (float) (2.0 * Math.PI * rate / sampleRate);
        final float twoPi = (float) (2.0 * Math.PI);
        final float halfDepth = depth * 0.5f;
        float ph = phase;
        final int n = input.length;
        for (int i = 0; i < n; i++) {
            float lfo = 1f - halfDepth * (1f - (float) Math.cos(ph));
            output[i] = input[i] * lfo;
            ph += phaseInc;
            if (ph > twoPi) ph -= twoPi;
        }
        phase = ph;
    }
}
